errors.go

```diff
@@ -0,0 +1,160 @@
+
+type prePostWrappedNotFoundError struct {
+	pre  string
+	post string
+
+	wrapped ipld.ErrNotFound
+}
+
+func (e prePostWrappedNotFoundError) String() string {
+	return e.Error()
+}
+
+func (e prePostWrappedNotFoundError) Error() string {
+	return e.pre + e.wrapped.Error() + e.post
+}
+
+func (e prePostWrappedNotFoundError) Unwrap() error {
+	return e.wrapped
+}
+
+func parseErrNotFoundWithFallbackToMSG(msg string) error {
+	err, handled := parseErrNotFound(msg)
+	if handled {
+		return err
+	}
+
+	return errors.New(msg)
+}
+
+func parseErrNotFoundWithFallbackToError(msg error) error {
+	err, handled := parseErrNotFound(msg.Error())
+	if handled {
+		return err
+	}
+
+	return msg
+}
+
+// Use a string to move it into RODATA
+// print("".join("\\x01" if chr(i) not in string.ascii_letters + string.digits else "\\x00" for i in range(ord('z')+1)))
+const notAsciiLetterOrDigitsLUT = "\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+
+func notAsciiLetterOrDigits(r rune) bool {
+	if r > 'z' {
+		return true
+	}
+
+	return notAsciiLetterOrDigitsLUT[r] > 0
+}
+
+//lint:ignore ST1008 this function is not using the error as a mean to return failure but it massages it to return the correct type
+func parseErrNotFound(msg string) (error, bool) {
+	if msg == "" {
+		return nil, true // Fast path
+	}
+
+	if err, handled := parseIPLDErrNotFound(msg); handled {
+		return err, true
+	}
+
+	if err, handled := parseBlockstoreNotFound(msg); handled {
+		return err, true
+	}
+
+	return nil, false
+}
+
+//lint:ignore ST1008 using error as values
+func parseIPLDErrNotFound(msg string) (error, bool) {
+	// The patern we search for is:
+	const ipldErrNotFoundKey = "ipld: could not find " /*CID*/
+	// We try to parse the CID, if it's invalid we give up and return a simple text error.
+	// We also accept "node" in place of the CID because that means it's an Undefined CID.
+
+	keyIndex := strings.Index(msg, ipldErrNotFoundKey)
+
+	if keyIndex < 0 { // Unknown error
+		return nil, false
+	}
+
+	cidStart := keyIndex + len(ipldErrNotFoundKey)
+
+	msgPostKey := msg[cidStart:]
+	var c cid.Cid
+	var postIndex int
+	if strings.HasPrefix(msgPostKey, "node") {
+		// Fallback case
+		c = cid.Undef
+		postIndex = len("node")
+	} else {
+		// Assume that CIDs only contain a-zA-Z0-9 characters.
+		// This is true because go-ipld-format use go-cid#Cid.String which use base{3{2,6},58}.
+		postIndex = strings.IndexFunc(msgPostKey, notAsciiLetterOrDigits)
+		if postIndex < 0 {
+			postIndex = len(msgPostKey)
+		}
+
+		var err error
```

Comment:
Do we really need the complexity brought by this? Can't we just assume the CID is separated by a space, instead? (if we have to assume anything at all)

Reply:
Just a space seems incomplete, I would like to include `"` too.

And at this point instead of searching a list of things I want to break on, I thought it was simpler to do a list of things I do not want to break on.

That is very static code that should never need an update (unless this parsing code is removed) and has a ~100% test coverage. So it's not really adding maintainment costs.

Reply:
> this code could be made simpler by removing the LUT
Yes, that.
> however not breaking on a-zA-Z0-9 seems usefull to me.
I'm not sure how this keeps something from not breaking, will need to take a closer look. In general hard-coding a basic alphanumeric detector along with its tests sounds like unneeded complexity added, if not for maintaining costs (though more code seems to always imply that) at least for my selfish reviewing costs.

Reply:
> I'm not sure how this keeps something from not breaking

I don't meant breaking code, I meant breaking as control flow, breaking out of the `IndexFunc` loop.

Since you are two with gus raising the same point https://github.com/ipfs/go-ipfs-http-client/pull/156#discussion_r840835826 I've replaced it by breaking on "` \t\n\r\v\f;\"`" (whitespaces or semicolon (produced by multierr when concatenating) or `"` for `%q` and `%#v`).